### DISCUSSION TITLE
Linux equivalent ?

### DISCUSSION THREAD

**Discussion by KaKi87** [ImranR98/Obtainium#1053] [General]
Hi,

This is not an implementation request, just a question : has anyone ever heard of any Linux equivalent to Obtainium, i.e. an app automatically downloads binaries/DEBs/RPMs for apps that aren't available in any repository or store ?

Thanks

**Comment by ImranR98** [comment]
Someone had mentioned https://github.com/marcosnils/bin in some issue

Edit: Mentioned in #777

**Reply by roke-julian-lockhart** [reply]
> [`#discussioncomment-7394610`](https://github.com/ImranR98/Obtainium/discussions/1053#discussioncomment-7394610)

@ImranR98, [`bin/discussions/182`](https://github.com/marcosnils/bin/discussions/182#discussion-5990206):

> I'd appreciate a Gui very much. Perhaps something similar to [Obtainium](https://github.com/ImranR98/Obtainium) for Android?

…and its sole official response, [`#discussioncomment-7945744`](https://github.com/marcosnils/bin/discussions/182#discussioncomment-7945744):

> Unfortunately, I personally don't have the time or the need to build a GUI. If you or someone feels like giving a try, happy to provide some support and feedback during the process.

…are noteworthy in this context.

**Comment by Mr-Bajs** [comment]
[deb-get](https://github.com/wimpysworld/deb-get)
Gets the latest deb's from websites.

**Reply by roke-julian-lockhart** [reply]
> [`deb-get`](https://github.com/wimpysworld/deb-get) gets the latest `deb`s from websites.

@Mr-Bajs, solely for Debian-based OSes. That's a significant amount of OSes, but it's also a significant drawback. The rationale for the limitation is evident, but it's not quite a replacement - imagine if Obtanium solely worked on Samsung's AOSP distribution.

**Reply by tranquil-tr0** [reply]
The other OSes also don't share packaging formats though, no? It supports debs in the way Obtanium supports apks, though I don't like it for different reasons (curated thus limited list).

**Reply by roke-julian-lockhart** [reply]
> [`#discussioncomment-13502309`](https://github.com/ImranR98/Obtainium/discussions/1053#discussioncomment-13502309)

@tranquil-tr0, that's a good point: It's not as if Obtanium downloads compiled sources and runs them from a directory like a portable installation. It installs the installation packages.

**Comment by summoner001** [comment]
Use Arch Linux, or an alternative (EndeavourOS), and enable [AUR](https://wiki.archlinux.org/title/Arch_User_Repository) (Arch User Repository) support. And use an AUR helper like paru or yay.

And what is not in the pacman/extras repo may be in the AUR, if it is not there then the git-clone command is there, if the three above fail then the program does not exist.

**Reply by roke-julian-lockhart** [reply]
> Use Arch Linux, or an alternative (EndeavourOS), and enable [AUR](https://wiki.archlinux.org/title/Arch_User_Repository) (Arch User Repository) support. And use an AUR helper like paru or yay.

@summoner001, that's at best equivalent to Fedora's COPR or openSUSE's OBS. It's not a way for individual users to automatically acquire updates to packages not yet uploaded to any package management repository via a local GUI-based application.

**Reply by summoner001** [reply]
AUR is equivalent with Obtainium. Pamac (package manager with GUI) can handle AUR packages on arch linux (and on Endeavour OS). But yes. If the package isn't available in AUR than use git... and you can build it for yourself.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/ImranR98/Obtainium/discussions/1053#discussioncomment-11019615

@summoner001, compiling sources provided by `git` isn't vaguely equivalent to what Obtanium provides. This is a discussion of how to automatically replace downloaded *binaries* from source-control websites' repositories' releases (which are not exposed to `git`).

What you're describing is equivalent to:

1.	Downloading a repository's sources via `git`;
1.	Compiling those with Android Studio; and
1.	Installing them via `pm` (or the `adb install` wrapper).

**Reply by summoner001** [reply]
Let's say then that for you there is no alternative to Obtainium. I have offered an alternative AUR+Pamac(GUI). AUR has the most packages. There are very few applications that are not there. Or very specific. If there is no package in AUR and it is difficult to use git, what do you recommend?

**Reply by roke-julian-lockhart** [reply]
> https://github.com/ImranR98/Obtainium/discussions/1053#discussioncomment-11019864

@summoner001, https://github.com/ImranR98/Obtainium/discussions/1053#discussioncomment-7394610 recommends something that is very close, for GitHub repositories. For non-GitHub repositories, https://github.com/ImranR98/Obtainium/discussions/1053#discussioncomment-11019633 is closest. Otherwise, your suggestion is as good as any other.

**Comment by KaKi87** [comment]
Hi everyone,

Since then, I started building my own tool named `dynapt` as in *DYNamic APT reposiory*, which auto-downloads DEB and AppImage files, and will in the future handle binaries.

Since it serves a local APT reposiory, it works with `apt` as well as any existing CLI and GUI supporting APT.

It's still way experimental though, you can try it by following [this tutorial](https://gist.github.com/KaKi87/f620788e9901abbfef4978eb7ad358b4).

**Comment by tranquil-tr0** [comment]
Hi all,

I've also made my own tool since I unfortunately never found any that worked for me. Try it out [here](https://github.com/tranquil-tr0/get). It (should) support any github repo publishing .deb files in releases (feel free to let me know if it does not). This supports any system that supports dpkg (which is what apt uses).

The idea was for it to be as simple as possible: `get install <user>/<repo>`

I'll be expanding on this rudimentary version, and I'd love to hear from anyone else who was looking for an Obtanium for Debian!

**Reply by KaKi87** [reply]
Hi,

> I unfortunately never found any that worked for me.

Does that include dynapt ?

Since you're using a DEB-based distro, I did mean for this tool to fit your needs as much as mine, so I would be interested in knowing how it doesn't so I could fix that.

Thank you

**Reply by tranquil-tr0** [reply]
Yeah for sure! The reason I personally didn't end up using dynapt was because I was looking for something super simple. I am not opposed to a more involved system, per se (I am maintaining a pacstall package), but I did want to avoid config files. I think another reason was because I was somewhat intimidated by your guide, especially since I last read it a while ago, when I was both less proficient and when dynapt was presumably less developed. I did just take a look at the guide again though and I am definitely considering setting up dynapt if I have something like beeper again (although I use it less now and it has update checking, so maybe not right now).
Edit: beeper *got* update checking as a feature recently I think

**Reply by roke-julian-lockhart** [reply]
> I've also made my own tool since I unfortunately never found any that worked for me.

@tranquil-tr0, why was `deb-get` unsitable? [^1]

[^1]: [`#discussioncomment-13502309`](https://github.com/ImranR98/Obtainium/discussions/1053#discussioncomment-13502309)

**Reply by tranquil-tr0** [reply]
Yeah, so the reason for me was that it seems deb-get had a limited repo, which hurt its feasibility since a part of the reasoning for me to use a Obtanium style package manager is to install super niche packages.

**Reply by roke-julian-lockhart** [reply]
> [`#discussioncomment-13508522`](https://github.com/ImranR98/Obtainium/discussions/1053#discussioncomment-13508522)

@tranquil-tr0, I'd say that makes it not Obtanium-equivalent, since the whole point of Obtanium is that there's no repository to maintain. Thanks.

**Reply by KaKi87** [reply]
> I did want to avoid config files

Yeah I'm planning on providing a CLI and a GUI, but I must admit I likely won't really get motivated until I get *some* early adoption 😅

**Reply by tranquil-tr0** [reply]
> > [`#discussioncomment-13508522`](https://github.com/ImranR98/Obtainium/discussions/1053#discussioncomment-13508522)
> 
> @tranquil-tr0, I'd say that makes it not Obtanium-equivalent, since the whole point of Obtanium is that there's no repository to maintain. Thanks.

Yeah, I mostly love Obtanium because of that, which is why deb-get didn't quite work for me (assuming that's what you were responding about).

**Comment by M-L-Ml** [comment]
For some python repos you can also do
https://www.geeksforgeeks.org/git/how-to-pip-install-from-a-git-repo-branch/#method-2-direct-installation-from-a-git-repository
but seems it would not support easy updates

**Reply by tranquil-tr0** [reply]
This seems like a great choice for python packages!

**Reply by KaKi87** [reply]
Especially with `uv` as a `pip` replacement 👍

**Comment by summoner001** [comment]
https://github.com/PlebOne/autonomix

Autonomix is an Obtainium like package manager for linux under development in Rust.

**Reply by roke-julian-lockhart** [reply]
@summoner001, thank you! Posted to [`softwarerecs.stackexchange.com/revisions/95043/1`](https://softwarerecs.stackexchange.com/revisions/95043/1#:~:text=For%20the%20majority%20of%20Linux%2Dbased%20OSes%2C%20github.com/PlebOne/autonomix%20is%20a%20near%2D1:1%20equivalent:%20%5B1%5D).

**Reply by summoner001** [reply]
Autonomix down :( RIP

After Autonomix rewritted in flutter, the repo goes down.

It seems this is a vibecoded garbage... so sorry :(

**Comment by shuvashish76** [comment]
For AppImages and portable apps : https://github.com/ivan-hc/AM
Especially see `am -t` command https://github.com/ivan-hc/AM/blob/main/docs/guides-and-tutorials/template.md for your own scripts.

**Reply by KaKi87** [reply]
Actually, I ended up creating [my own tool](https://gist.github.com/KaKi87/f620788e9901abbfef4978eb7ad358b4) that packages AppImage apps into DEBs and serves them to APT directly so I don't have to use a separate CLI to update them : APT will do it at the same time as other apps.

**Reply by shuvashish76** [reply]
Experimental GUI available for AM https://github.com/Shikakiben/AM-GUI

**Reply by roke-julian-lockhart** [reply]
@shuvashish76, does it support installable packages of any kind, or solely portable package types?

**Reply by shuvashish76** [reply]
I'm not sure about the GUI version but you can use AM package manager as root or rootless.

> installable packages of any kind, or solely portable package types?

Isn't installable packages = binaries stored in root location for safety and executable and having `.desktop` file?
You are kinda doing same with AM package-manager, they're stored in https://github.com/ivan-hc/AM?tab=readme-ov-file#how-are-apps-installed , https://github.com/ivan-hc/AM?tab=readme-ov-file#how-to-set-the-path-to-local-apps and does extract/add `.desktop` files for you.

But yes AM is package manager for specifically AppImages and Portable apps that supposed to work on any linux.

**Reply by roke-julian-lockhart** [reply]
> Isn't installable packages = binaries stored in root location

@shuvashish76, when I wrote that, I was under the impression that, because the FHS is very complex, many of the installable-to locations are non-PID0-RW, with `/opt` as an example. However, `stat /opt --format=%G` and [`stackoverflow.com/revisions/74132011/1`](https://stackoverflow.com/revisions/74132011/1#content:~:text=/opt%20%2D%20Designed%20for%20non%2DOS%20packages%2E) demonstrate otherwise.

> AM is package manager for specifically AppImages and Portable apps

What I was asking was whether it can, in-effect, install an `.RPM` or `.Deb`, etcetera. If it can extract its content to `/opt`, then add the `.desktop` file to `$XDG_DATA_DIRS`, I presume that that should be sufficient. I ask because Obtanium obviously can install `.APKs`.
